### ISSUE TITLE
can not reproduce snapshot ensemble

### ISSUE THREAD

**Issue by junbum766** [nannullna/snapshot-al#15]
I have tried the SE method several times but cannot reproduce it.
Below are the reproduction results of SE applying the maxentropy acquisition function.
I haven't modified any config files or other code.
Also I wonder Table 1 in the paper, 'DE at the final episode' result is same as 'ens acc'?

> Before SWA 0.319 1.990 0.196
> Episode 1 num_models: 5 -- max eval acc: 35.60, test acc: 33.15
> model_0 test accuracy: 30.17% model_1 test accuracy: 31.53% model_2 test accuracy: 31.09% model_3 test accuracy: 33.09% model_4 test accuracy: 34.69% 
> ens acc: 34.75, mean acc: 32.11
> Before SWA 0.383 2.083 0.278
> Episode 2 num_models: 5 -- max eval acc: 42.20, test acc: 37.72
> model_0 test accuracy: 40.65% model_1 test accuracy: 40.73% model_2 test accuracy: 40.16% model_3 test accuracy: 39.58% model_4 test accuracy: 37.38% 
> ens acc: 42.16, mean acc: 39.70
> Before SWA 0.451 1.830 0.249
> Episode 3 num_models: 5 -- max eval acc: 53.40, test acc: 49.29
> model_0 test accuracy: 48.92% model_1 test accuracy: 17.72% model_2 test accuracy: 50.52% model_3 test accuracy: 50.62% model_4 test accuracy: 17.23% 
> ens acc: 47.23, mean acc: 37.00
> Before SWA 0.435 1.895 0.274
> Episode 4 num_models: 5 -- max eval acc: 56.60, test acc: 53.00
> model_0 test accuracy: 51.07% model_1 test accuracy: 42.14% model_2 test accuracy: 52.83% model_3 test accuracy: 50.60% model_4 test accuracy: 51.09% 
> ens acc: 55.21, mean acc: 49.55
> Before SWA 0.562 1.344 0.145
> Episode 5 num_models: 5 -- max eval acc: 62.80, test acc: 61.88
> model_0 test accuracy: 54.95% model_1 test accuracy: 52.80% model_2 test accuracy: 51.67% model_3 test accuracy: 61.21% model_4 test accuracy: 43.83% 
> ens acc: 61.91, mean acc: 52.89
> Before SWA 0.576 1.284 0.105
> Episode 6 num_models: 5 -- max eval acc: 64.20, test acc: 64.00
> model_0 test accuracy: 41.49% model_1 test accuracy: 53.28% model_2 test accuracy: 61.31% model_3 test accuracy: 61.69% model_4 test accuracy: 56.28% 
> ens acc: 64.60, mean acc: 54.81
> Before SWA 0.498 1.673 0.238
> Episode 7 num_models: 5 -- max eval acc: 60.80, test acc: 63.57
> model_0 test accuracy: 50.48% model_1 test accuracy: 53.26% model_2 test accuracy: 32.01% model_3 test accuracy: 59.14% model_4 test accuracy: 53.27% 
> ens acc: 62.22, mean acc: 49.63
> Before SWA 0.556 1.448 0.170
> Episode 8 num_models: 5 -- max eval acc: 60.80, test acc: 70.98
> model_0 test accuracy: 20.22% model_1 test accuracy: 59.47% model_2 test accuracy: 60.10% model_3 test accuracy: 56.27% model_4 test accuracy: 63.02% 
> ens acc: 68.13, mean acc: 51.82
> Before SWA 0.489 1.753 0.250
> Episode 9 num_models: 5 -- max eval acc: 64.60, test acc: 74.96
> model_0 test accuracy: 58.13% model_1 test accuracy: 58.42% model_2 test accuracy: 61.01% model_3 test accuracy: 61.25% model_4 test accuracy: 60.08% 
> ens acc: 70.70, mean acc: 59.78
> Before SWA 0.454 2.085 0.310
> Episode 10 num_models: 5 -- max eval acc: 61.20, test acc: 72.86
> model_0 test accuracy: 52.16% model_1 test accuracy: 51.13% model_2 test accuracy: 55.79% model_3 test accuracy: 57.12% model_4 test accuracy: 61.02% 
> ens acc: 68.43, mean acc: 55.44
> Before SWA 0.540 1.551 0.217
> Episode 11 num_models: 5 -- max eval acc: 60.40, test acc: 73.67
> model_0 test accuracy: 60.16% model_1 test accuracy: 52.73% model_2 test accuracy: 58.61% model_3 test accuracy: 58.09% model_4 test accuracy: 59.18% 
> ens acc: 68.75, mean acc: 57.75
> Before SWA 0.593 1.350 0.174
> Episode 12 num_models: 5 -- max eval acc: 66.20, test acc: 77.59
> model_0 test accuracy: 62.23% model_1 test accuracy: 61.37% model_2 test accuracy: 58.18% model_3 test accuracy: 63.58% model_4 test accuracy: 62.04% 
> ens acc: 72.37, mean acc: 61.48
> Before SWA 0.594 1.408 0.191
> Episode 13 num_models: 5 -- max eval acc: 67.40, test acc: 78.78
> model_0 test accuracy: 57.97% model_1 test accuracy: 56.17% model_2 test accuracy: 64.93% model_3 test accuracy: 63.59% model_4 test accuracy: 61.30% 
> ens acc: 73.23, mean acc: 60.79
> Before SWA 0.608 1.333 0.169
> Episode 14 num_models: 5 -- max eval acc: 70.20, test acc: 80.49
> model_0 test accuracy: 68.08% model_1 test accuracy: 56.76% model_2 test accuracy: 64.01% model_3 test accuracy: 63.36% model_4 test accuracy: 66.73% 
> ens acc: 76.79, mean acc: 63.79
> Before SWA 0.612 1.335 0.184
> Episode 15 num_models: 5 -- max eval acc: 69.20, test acc: 81.86
> model_0 test accuracy: 60.53% model_1 test accuracy: 61.63% model_2 test accuracy: 63.45% model_3 test accuracy: 68.01% model_4 test accuracy: 65.90% 
> ens acc: 76.15, mean acc: 63.90
> Before SWA 0.650 1.176 0.147
> Episode 16 num_models: 5 -- max eval acc: 70.40, test acc: 82.81
> model_0 test accuracy: 67.36% model_1 test accuracy: 66.95% model_2 test accuracy: 64.89% model_3 test accuracy: 67.66% model_4 test accuracy: 68.88% 
> ens acc: 79.13, mean acc: 67.15
> Before SWA 0.672 1.053 0.125
> Episode 17 num_models: 5 -- max eval acc: 71.00, test acc: 83.91
> model_0 test accuracy: 64.46% model_1 test accuracy: 66.32% model_2 test accuracy: 68.62% model_3 test accuracy: 65.99% model_4 test accuracy: 70.04% 
> ens acc: 77.79, mean acc: 67.09
> Before SWA 0.597 1.439 0.216
> Episode 18 num_models: 5 -- max eval acc: 73.60, test acc: 84.93
> model_0 test accuracy: 69.62% model_1 test accuracy: 70.47% model_2 test accuracy: 68.63% model_3 test accuracy: 65.17% model_4 test accuracy: 67.21% 
> ens acc: 80.40, mean acc: 68.22
> Before SWA 0.692 0.994 0.113
> Episode 19 num_models: 5 -- max eval acc: 75.40, test acc: 85.09
> model_0 test accuracy: 66.88% model_1 test accuracy: 68.20% model_2 test accuracy: 69.52% model_3 test accuracy: 69.78% model_4 test accuracy: 71.86% 
> ens acc: 81.33, mean acc: 69.25
> Before SWA 0.589 1.486 0.210
> Episode 20 num_models: 5 -- max eval acc: 74.40, test acc: 86.23
> model_0 test accuracy: 71.40% model_1 test accuracy: 68.54% model_2 test accuracy: 71.38% model_3 test accuracy: 68.60% model_4 test accuracy: 68.02% 
> ens acc: 82.00, mean acc: 69.59
> Before SWA 0.675 1.114 0.151
> Episode 21 num_models: 5 -- max eval acc: 73.60, test acc: 86.40
> model_0 test accuracy: 70.47% model_1 test accuracy: 64.42% model_2 test accuracy: 72.41% model_3 test accuracy: 60.67% model_4 test accuracy: 67.00% 
> ens acc: 81.10, mean acc: 66.99
> Before SWA 0.720 0.892 0.101
> Episode 22 num_models: 5 -- max eval acc: 75.00, test acc: 87.08
> model_0 test accuracy: 71.12% model_1 test accuracy: 70.54% model_2 test accuracy: 73.33% model_3 test accuracy: 71.64% model_4 test accuracy: 71.30% 
> ens acc: 82.85, mean acc: 71.59
> Before SWA 0.697 0.975 0.121
> Episode 23 num_models: 5 -- max eval acc: 74.00, test acc: 87.31
> model_0 test accuracy: 71.39% model_1 test accuracy: 65.91% model_2 test accuracy: 70.73% model_3 test accuracy: 68.48% model_4 test accuracy: 71.07% 
> ens acc: 80.63, mean acc: 69.52
> Before SWA 0.735 0.831 0.077
> Episode 24 num_models: 5 -- max eval acc: 76.60, test acc: 88.21
> model_0 test accuracy: 72.81% model_1 test accuracy: 69.62% model_2 test accuracy: 73.12% model_3 test accuracy: 74.81% model_4 test accuracy: 73.01% 
> ens acc: 84.65, mean acc: 72.67
> Before SWA 0.710 0.917 0.096
> Episode 25 num_models: 5 -- max eval acc: 78.60, test acc: 88.40
> model_0 test accuracy: 73.07% model_1 test accuracy: 76.46% model_2 test accuracy: 71.14% model_3 test accuracy: 77.80% model_4 test accuracy: 72.12% 
> ens acc: 84.28, mean acc: 74.12

**Comment by nannullna** [comment]
Thank you for asking questions on our paper and the code. 

`ens acc` are not the reported figures of "DE at the final episode" in the paper. The reported figures are the results of re-training an ensemble of five members with those acquired samples.

For SE to work well in the context of AL, we observed that the diversity of predictions is important. Therefore, for experiments on CIFAR-10 dataset, we increased the learning rate while gathering snapshots as explained in the Algorithm 1 and Appendix. Therefore, the accuracy of each snapshot might be lower, which may not be suitable for true performance evaluation.

You can get the results in the paper by training a new model, or an ensemble of models with labeled samples after the AL process. For re-training, we basically used the same hyperparameters as ones used in the active learning process. 

As we highly emphasized in the section 3.2 by saying _"what really matters is the accuracy after the final episode,"_ we disentangled (or at least tried to disentangle) the quality of acquired examples from the training method for evaluation. If you manage to select good examples to be labeled, the performance of a model trained with them should be higher regardless of training methods (a single model, DE, or even SWA) or hyperparameters, which is one of the reasons we decided to include DE results in the tables. 

Feel free to ask me if you have more questions.

**Comment by junbum766** [comment]
As you commented, I retrained the DE model from scratch with data from the SE method and it is now reproducible.
thank you!
